crates/fluvio-protocol/src/core/bytebuf.rs

```diff
@@ -0,0 +1,262 @@
+use std::io::Error;
+use std::io::ErrorKind;
+use std::ops::Deref;
+
+use bytes::{Buf, Bytes, BufMut};
+
+use crate::{Encoder, Decoder, Version};
+
+/// Represnts a SmartModule WASM File bytes.
+///
+/// Provides a `Encoder` and `Decoder` implementation optimized for WASM files
+/// used in SmartModules.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct ByteBuf {
+    inner: Bytes,
+}
+
+impl From<Vec<u8>> for ByteBuf {
+    fn from(bytes: Vec<u8>) -> Self {
+        ByteBuf {
+            inner: Bytes::from_iter(bytes.into_iter()),
+        }
+    }
+}
+
+impl From<ByteBuf> for Vec<u8> {
+    fn from(bytebuf: ByteBuf) -> Self {
+        bytebuf.inner.to_vec()
+    }
+}
+
+impl Deref for ByteBuf {
+    type Target = Bytes;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl Decoder for ByteBuf {
+    fn decode<T>(&mut self, src: &mut T, version: Version) -> Result<(), Error>
+    where
+        T: Buf,
+    {
+        let mut len: u32 = 0;
+        len.decode(src, version)?;
+
+        if len < 1 {
+            return Ok(());
+        }
+
+        self.inner = src.copy_to_bytes(len as usize);
+
+        Ok(())
+    }
+}
+
+impl Encoder for ByteBuf {
+    fn write_size(&self, version: Version) -> usize {
+        0_u32.write_size(version) + self.inner.len()
+    }
+
+    fn encode<T>(&self, dest: &mut T, version: Version) -> Result<(), Error>
+    where
+        T: BufMut,
+    {
+        let remaining = dest.remaining_mut();
+        let expected = self.write_size(version);
+
+        if remaining < expected {
+            return Err(Error::new(
+                ErrorKind::UnexpectedEof,
+                format!(
+                    "Not enough capacity for ByteBuf. Expected: {}, Remaining: {}",
+                    expected, remaining
+                ),
+            ));
+        }
+
+        dest.put_u32(self.inner.len() as u32);
+        dest.put(self.inner.clone());
+
+        Ok(())
+    }
+
+    fn as_bytes(&self, _version: Version) -> Result<Bytes, Error> {
+        Ok(self.inner.clone())
+    }
+}
+
+#[cfg(feature = "use_serde")]
```

Comment:
I don't think ths belongs to this create.  This is opinonated encoding which can be moved back to controlplane protocol

Reply:
Addressed on: [630aa06](https://github.com/infinyon/fluvio/pull/2738/commits/630aa06d16fae114d94efae22924cd9060650394)